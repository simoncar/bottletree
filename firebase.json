{
  "emulators": {
    "auth": {
      "port": 9099
    },
    "firestore": {
      "port": 8080
    },
    "ui": {
      "enabled": true
    },
    "singleProjectMode": true
  },
  "firestore": {
    "rules": "firestore.rules",
    "indexes": "firestore.indexes.json"
  },
  "storage": {
    "rules": "storage.rules"
  },
  "functions": [
    {
      "source": "functions",
      "codebase": "default",
      "ignore": [
        "node_modules",
        ".git",
        "firebase-debug.log",
        "firebase-debug.*.log"
      ]
    }
  ]
}
