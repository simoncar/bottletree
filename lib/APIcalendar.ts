import { addDoc, collection, onSnapshot, query } from "firebase/firestore";
import { db } from "./firebase";
import { IProject, ICalendarEvent } from "./types";

type projectsRead = (projects: IProject[]) => void;

export async function getItems(projectId: string, callback: itemsRead) {
    const q = query(collection(db, "calendar"));

    const unsubscribe = onSnapshot(q, (querySnapshot) => {
        const calendarEvents: ICalendarEvent[] = [];
        querySnapshot.forEach((doc) => {
            calendarEvents.push({
                key: doc.id,
                allDay: doc.data().allDay,
                dateBegin: doc.data().dateBegin,
                dateEnd: doc.data().dateEnd,
                description: doc.data().description,
                projectId: doc.data().projectId,
                title: doc.data().title,
                uid: doc.data().uid,
                dateBeginSplit: doc
                    .data()
                    .dateBegin.toDate()
                    .toISOString()
                    .split("T")[0],
                dateEndSplit: doc
                    .data()
                    .dateEnd.toDate()
                    .toISOString()
                    .split("T")[0],
                timeBegin: doc
                    .data()
                    .dateBegin.toDate()
                    .toLocaleTimeString("en-US"),
                timeEnd: doc
                    .data()
                    .dateEnd.toDate()
                    .toLocaleTimeString("en-US"),
            });
        });

        const expandedDates = expandDateRanges(calendarEvents);
        console.log("expandedDates", expandedDates);

        callback(expandedDates);
    });

    return () => unsubscribe();
}

interface DateRange {
    dateBegin: string;
    dateEnd: string;
}

interface ExpandedDates {
    [date: string]: string;
}

function expandDateRanges(dateRanges: DateRange[]): ExpandedDates {
    const expandedDates: ExpandedDates = {};

    for (const range of dateRanges) {
        const startDate = new Date(range.dateBeginSplit);
        const endDate = new Date(range.dateEndSplit);

        const currentDate = new Date(startDate);
        while (currentDate <= endDate) {
            const dateString = currentDate.toISOString().split("T")[0];
            expandedDates[dateString] = [
                ...(expandedDates[dateString] || []),
                {
                    title: range.title,
                    description: range.description,
                    projectId: range.projectId,
                    key: range.key,
                    dateBegin: range.dateBegin,
                    dateEnd: range.dateEnd,
                    allDay: range.allDay,
                    uid: range.uid,
                    dateBeginSplit: range.dateBeginSplit,
                    timeBegin: range.timeBegin,
                    timeEnd: range.timeEnd,
                },
            ];

            currentDate.setDate(currentDate.getDate() + 1);
        }
    }

    return expandedDates;
}

export function addCalendarEvent(
    calendarEvent: ICalendarEvent,
    callback: { (id: string): void; (arg0: string): void },
) {
    try {
        addDoc(collection(db, "calendar"), {
            allDay: calendarEvent.allDay,
            dateBegin: calendarEvent.dateBegin,
            dateEnd: calendarEvent.dateEnd,
            description: calendarEvent.description,
            projectId: calendarEvent.projectId,
            title: calendarEvent.title,
            uid: calendarEvent.uid,
        }).then((docRef) => {
            callback(docRef.id);
        });
    } catch (e) {
        console.error("Error adding calendar event: ", e);
    }

    return;
}
