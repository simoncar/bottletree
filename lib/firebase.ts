import * as Device from "expo-device";
import firebase from "@react-native-firebase/app";
import auth from "@react-native-firebase/auth";
import firestore from "@react-native-firebase/firestore";
import storage from "@react-native-firebase/storage";

export const firebaseErrors: Record<string, string> = {
  "Firebase: Error (auth/email-already-in-use).":
    "This email is already is use.",
  "Firebase: Error (auth/popup-closed-by-user).":
    "Process got canceled by user.",
  "Firebase: Error (auth/user-not-found).": "User not found.",
  "Firebase: Error (auth/wrong-password).": "Wrong email or password.",
  "Firebase: Error (auth/network-request-failed).":
    "Failed to send the request. Check your connection.",
};

const db = firestore();

console.log("__DEV__: ", __DEV__);

if (!Device.isDevice) {
  console.log("Connecting to Firebase Emulator");
  db.useEmulator("127.0.0.1", 8080);
  firestore().useEmulator("localhost", 8080);
  auth().useEmulator("http://localhost:9099");
  storage().useEmulator("127.0.0.1", 9199);
}

export { firebase, db, storage, auth, firestore };
