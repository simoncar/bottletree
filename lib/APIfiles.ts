import { dbm } from "@/lib/firebase"; // or "@/lib/firebase.web"
import {
  addDoc,
  collection,
  doc,
  onSnapshot,
  orderBy,
  query,
  serverTimestamp,
} from "@react-native-firebase/firestore";

// import {
//   getDownloadURL,
//   getStorage,
//   ref as storageRef,
// } from "@react-native-firebase/storage";
import storage, { getDownloadURL } from "@react-native-firebase/storage";

import * as Crypto from "expo-crypto";
import * as DocumentPicker from "expo-document-picker";
import { setPostFile } from "./APIpost";
import { IFile, IPost, IUser } from "./types";

type FileData = {
  mimeType: string;
  name: string;
  size: number;
  uri: string;
};

type FilesJson = {
  assets: FileData[];
  canceled: boolean;
};

export async function deleteFile(
  projectId: string,
  key: string,
  callback: (id: string) => void,
) {
  try {
    const docRef = doc(
      collection(doc(collection(dbm, "projects"), projectId), "files"),
      key,
    );

    await dbm.runTransaction(async (transaction) => {
      transaction.delete(docRef);
    });

    callback(key);
  } catch (e) {
    console.error("Error deleting file from project: ", e);
  }
}
export async function getFiles(project: string, callback: filesRead) {
  const filesRef = collection(
    doc(collection(dbm, "projects"), project),
    "files",
  );
  const q = query(filesRef, orderBy("modified", "desc"));

  const unsubscribe = onSnapshot(q, (querySnapshot) => {
    const files: IFile[] = [];

    querySnapshot.forEach((docSnap) => {
      const data = docSnap.data();
      files.push({
        key: docSnap.id,
        filename: data.filename,
        url: data.url,
        mimeType: data.mimeType,
        bytes: data.bytes,
        created: data.created,
        modified: data.modified,
        projectId: data.projectId,
      });
    });

    callback(files);
  });

  return unsubscribe;
}

export async function uploadFilesAndCreateEntries(
  filesJson: DocumentPicker.DocumentPickerResult,
  projectId: string,
  user: IUser,
) {
  if (filesJson.canceled) {
    console.log("File selection was canceled.");
    return;
  }

  const projectFilesRef = collection(doc(dbm, "projects", projectId), "files");
  console.log("uploadFilesAndCreateEntries - modular:", filesJson);

  for (const file of filesJson.assets) {
    try {
      const UUID = Crypto.randomUUID();
      //const blob = await uriToBlob(file.uri);

      const filePath = `project/${projectId}/files/${UUID}`;
      const fileStorageRef = storage().ref(filePath);

      try {
        console.log("AAA111- fileStorageRef.putFile");
        await fileStorageRef.putFile(file.uri);
        console.log("AAA222- fileStorageRef.putFile");
        //await uploadBytes(fileStorageRef, blob);
      } catch (error: any) {
        if (error.code === "storage/unauthorized") {
          console.error("User does not have permission to access the object");
        } else {
          throw error;
        }
      }

      const downloadURL = await getDownloadURL(fileStorageRef);

      const fileEntry: IFile = {
        key: UUID,
        filename: file.name,
        url: downloadURL,
        mimeType: file.mimeType,
        bytes: file.size,
        created: serverTimestamp(),
        modified: serverTimestamp(),
        projectId,
      };

      await addDoc(projectFilesRef, fileEntry);

      const post: IPost = {
        key: UUID,
        caption: `${file.name}`,
        projectId,
        author: user.displayName,
        uid: user.uid,
        timestamp: serverTimestamp(),
        ratio: 1,
        file: UUID,
      };

      setPostFile(post, (postId) => {
        console.log(`Post created with ID: ${postId}`);
      });

      console.log(`File ${file.name} uploaded and entry created in Firestore.`);
    } catch (error) {
      console.error(`Error uploading file ${file.name}:`, error);
    }
  }
}
