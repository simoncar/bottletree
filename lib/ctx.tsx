import { updateAccountName, updateUser } from "@/lib/APIuser";
import {
  auth,
  firestore,
  signInWithEmailAndPassword,
  createUserWithEmailAndPassword,
  signInAnonymously,
  sendPasswordResetEmail,
  signOut as firebaseSignOut,
} from "@/lib/firebase";
import { IUser } from "@/lib/types";
import AsyncStorage from "@react-native-async-storage/async-storage";
import React, { createContext, useContext } from "react";
import { useStorageState } from "./useStorageState";
import {
  addDoc,
  collection,
  deleteDoc,
  doc,
  getDoc,
  getDocs,
  onSnapshot,
  orderBy,
  query,
  serverTimestamp,
  setDoc,
} from "@react-native-firebase/firestore";

interface AuthContextType {
  session?: string | null;
  isAuthLoading: boolean;
  resetPassword: (email: string, callback: (message: string) => void) => void;
  deleteAccount: (callback: (message: string) => void) => void;
  signIn: (email: string, password: string) => void;
  signUp: (
    screenName: string,
    screenEmail: string,
    screenPassword: string,

    callback: (user: IUser, message: string) => void,
  ) => void;
  signOut: () => void;
  signInAnonymously: (callback: (user: IUser, message: string) => void) => void;
}

export const AuthContext = createContext<AuthContextType>({
  session: null,
  isAuthLoading: true,
  resetPassword: () => null,
  deleteAccount: () => null,
  signIn: () => null,
  signUp: () => null,
  signOut: () => null,
  signInAnonymously: () => null,
});

// This hook can be used to access the user info.
export function useSession() {
  const context = useContext(AuthContext);

  if (!context) {
    throw new Error("useSession must be used within a Provider");
  }
  return context;
}

export function SessionProvider(props: React.PropsWithChildren) {
  const [[isAuthLoading, session], setSession] = useStorageState("session");

  const handleSignInAnonymously = async (
    callback: (session, message: string) => void,
  ) => {
    try {
      const response = await signInAnonymously(auth);
      await setSession(response.user.uid);
      updateAccountName(response.user.uid, "");
      console.log(
        "Sign In  auth().signInAnonymously Success",
        response.user.uid,
      );
      callback(response.user.uid, "Success");
      return response.user.uid;
    } catch (error) {
      setSession(null);
      console.log(error);
      callback(null, error);
    }
  };

  const signIn = async (email: string, password: string) => {
    try {
      const response = await signInWithEmailAndPassword(auth, email, password);
      setSession(response.user.uid);
      //updateAccountName(response.user.uid, "");

      console.log("Sign In Success", session, response.user.uid);
    } catch (error) {
      setSession(null);
      console.log(error);
      throw error;
    }
  };

  const signUp = async (
    name: string,
    email: string,
    password: string,
    callback: (user: IUser, message: string) => void,
  ) => {
    console.log("signUp email: ", email);

    try {
      const response = await createUserWithEmailAndPassword(
        auth,
        email,
        password,
      );

      setSession(response.user.uid);

      const user: IUser = {
        uid: auth.currentUser.uid,
        email: email,
        displayName: convertToString(name),
        photoURL: convertToString(""),
        project: "",
        anonymous: auth.currentUser.isAnonymous,
        created: serverTimestamp(),
      };

      console.log("signUp user: ", user);
      updateUser(user);
      //setUser(user);
      //updateAccountName(response.user.uid, name);

      callback(user, "SignUp Success");
    } catch (error) {
      console.log("signUp Error: ", error);
      callback(null, error);
    }
  };

  const signOut = async () => {
    try {
      setSession(null);
      if (auth.currentUser) {
        await firebaseSignOut(auth);
      }
      await AsyncStorage.clear();
      console.log("AsyncStorage cleared");
    } catch (e) {
      console.error("Failed to clear Session, auth or AsyncStorage:", e);
    }
  };

  const resetPassword = (screenEmail: string, callback: any) => {
    sendPasswordResetEmail(auth, screenEmail)
      .then(() => callback("Password reset email sent.  Check your inbox."))
      .catch((error) => {
        const errorMessage = error.message;
        callback(errorMessage);
      });
    //setSharedDataUser(null);
    //removeItem();
  };

  const deleteAccount = async (callback: any) => {
    const user = auth.currentUser;

    user
      .delete()
      .then(() => callback(null))
      .catch((error) => {
        const errorMessage = error.message;
        console.log("deleteAccount failure: ", error);

        callback(errorMessage);
      });
  };

  function convertToString(value: string | null): string {
    if (value === null) {
      return "";
    } else {
      return value;
    }
  }

  return (
    <AuthContext.Provider
      value={{
        isAuthLoading,
        session,
        resetPassword,
        deleteAccount,
        signInAnonymously: handleSignInAnonymously,
        signIn,
        signUp,
        signOut,
      }}
    >
      {props.children}
    </AuthContext.Provider>
  );
}
