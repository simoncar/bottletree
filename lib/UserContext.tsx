import React, { createContext, useEffect, useState } from "react";
import { getUser } from "@/lib/APIuser";
import { useSession } from "@/lib/ctx";
import { IUser } from "@/lib/types";

interface UserContextType {
  user: IUser | null;
  setUser: (user: IUser) => void;
}

const UserContext = createContext<UserContextType>({
  user: null,
  setUser: () => null,
});

const UserProvider = (props: React.PropsWithChildren) => {
  const { session } = useSession();
  const [user, setUser] = useState(null);

  useEffect(() => {
    const fetchData = async () => {
      try {
        const userData = await getUser(session);

        setUser({ ...userData });
      } catch (error) {
        console.error("error fetching user data in UserProvider", error);
      }
    };

    fetchData();
  }, []);

  return (
    <UserContext.Provider value={{ user, setUser }}>
      {props.children}
    </UserContext.Provider>
  );
};

export { UserContext, UserProvider };
