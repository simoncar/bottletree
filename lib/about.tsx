import React, { useContext, useState, useEffect } from "react";
import { Button, Pressable, StyleSheet } from "react-native";
import { Text, View } from "@/components/Themed";
import { useSession } from "@/lib/ctx";
import { auth } from "@/lib/firebase";
import { UserContext } from "@/lib/UserContext";
import { Link } from "expo-router";
import * as Application from "expo-application";
import * as Sentry from "@sentry/react-native";
import { useRouter } from "expo-router";

export const About = () => {
  const { user } = useContext(UserContext);
  const { session } = useSession();
  const [authUser, setAuthUser] = useState<string>(null);

  useEffect(() => {
    setAuthUser(auth().currentUser?.uid);
  }, []);

  const router = useRouter();
  let pressCount = 0;

  const handlePress = () => {
    pressCount += 1;
    console.log(pressCount);

    if (pressCount === 3) {
      router.push("/p");
    }
  };

  return (
    <View style={styles.aboutContainer}>
      <Pressable onPress={handlePress}>
        <Text style={styles.version}>One Build</Text>
      </Pressable>

      <Text style={styles.version}>
        {Application.nativeApplicationVersion} ({Application.nativeBuildVersion}
        ) | 12.app.sentry
      </Text>
      <Text style={styles.version}>Email: {user?.email}</Text>
      <Text style={styles.version}>Name: {user?.displayName}</Text>
      <Text style={styles.version}>User Context - {user?.uid}</Text>
      <Text style={styles.version}>Session - {session}</Text>
      <Text style={styles.version}>Auth - {auth().currentUser?.uid}</Text>
    </View>
  );
};

const styles = StyleSheet.create({
  aboutContainer: {
    alignItems: "center",
    justifyContent: "center",
    paddingBottom: 50,
    paddingTop: 50,
  },
  version: {
    color: "grey",
    fontSize: 10,
  },
});
