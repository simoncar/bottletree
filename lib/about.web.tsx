import React, { useState, useEffect, useContext } from "react";
import { StyleSheet } from "react-native";
import { getLocales } from "expo-localization";
import { Text, View } from "@/components/Themed";
import * as Application from "expo-application";
import { useSession } from "@/lib/ctx";
import { auth } from "@/lib/firebase";
import { UserContext } from "@/lib/UserContext";

export const About = () => {
  const { user } = useContext(UserContext);
  const { session } = useSession();
  const [authUser, setAuthUser] = useState<string>(true);
  const deviceLanguage = getLocales()[0].languageCode;

  useEffect(() => {
    setAuthUser(auth().currentUser?.uid);
  }, []);

  if (null == user) {
    return <View></View>;
  } else {
    return (
      <View style={styles.aboutContainer}>
        <Text style={styles.version}>{user.displayName}</Text>
        <Text style={styles.version}>{user.email}</Text>
        <Text style={styles.version}>
          Anonymous: {auth().currentUser?.isAnonymous}
        </Text>
        <Text style={styles.version}>
          {Application.nativeApplicationVersion} (
          {Application.nativeBuildVersion}) | 2
        </Text>

        <Text style={styles.version}>User Context - {user.uid}</Text>
        <Text style={styles.version}>CTX - {session}</Text>
        <Text style={styles.version}>Auth2 - {authUser}</Text>
        <Text style={styles.version}>Language - {deviceLanguage}</Text>
        <Text style={styles.version}>Project - {user.project}</Text>
      </View>
    );
  }
};

const styles = StyleSheet.create({
  aboutContainer: {
    alignItems: "center",
    justifyContent: "center",
    paddingTop: 50,
  },
  version: {
    color: "grey",
    fontSize: 14,
  },
});
