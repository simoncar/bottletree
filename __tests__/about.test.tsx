import React from "react";
import { About } from "@/lib/about";
import { renderRouter, render, screen } from "expo-router/testing-library";
import { UserContext } from "@/lib/UserContext";
import { getLocales } from "expo-localization";
import { auth } from "@/lib/firebase";
import { getSession } from "@/lib/ctx";

// Mock expo-localization
jest.mock("expo-localization", () => ({
  getLocales: jest.fn().mockReturnValue([
    {
      country: "US",
      languageTag: "en-US",
      languageCode: "en",
      isRTL: false,
    },
  ]),
}));

// Mock Firebase Auth
jest.mock("@/lib/firebase", () => {
  const auth = {
    signInWithEmailAndPassword: jest.fn().mockResolvedValue({
      user: {
        uid: "1234567890",
        displayName: "John Doe",
        email: "john.doe@example.com",
        photoURL: "https://example.com/john-doe.jpg",
      },
    }),
    signOut: jest.fn().mockResolvedValue({}),
    onAuthStateChanged: jest.fn((callback) => {
      callback({
        uid: "1234567890",
        displayName: "John Doe",
        email: "john.doe@example.com",
        photoURL: "https://example.com/john-doe.jpg",
      });
    }),
  };
  return {
    auth: jest.fn(() => auth),
  };
});

// Mock session in @/lib/ctx
jest.mock("@/lib/ctx", () => ({
  useSession: jest.fn().mockResolvedValue({
    user: {
      uid: "1234567890",
      displayName: "John Doe",
      email: "john.doe@example.com",
      photoURL: "https://example.com/john-doe.jpg",
    },
    token: "mocked-token",
  }),
}));

it("about test", async () => {
  const user = {
    uid: "1234567890",
    displayName: "John Doe",
    email: "john.doe@example.com",
    photoURL: "https://example.com/john-doe.jpg",
    project: "73JwAXeOEhLXUggpVKK9",
  };

  const setUser = jest.fn();

  const { toJSON } = render(
    <UserContext.Provider value={{ user, setUser }}>
      <About />
    </UserContext.Provider>,
  );

  expect(toJSON()).toMatchSnapshot();
});
