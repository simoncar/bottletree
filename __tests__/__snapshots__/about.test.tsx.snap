// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`about test 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "justifyContent": "center",
      "paddingTop": 50,
    }
  }
>
  <Text
    href="/user/"
    onPress={[Function]}
    role="link"
  >
    <Text
      style={
        [
          {
            "color": "#000",
          },
          {
            "color": "grey",
            "fontSize": 14,
          },
        ]
      }
    >
      John Doe
    </Text>
  </Text>
  <Text
    style={
      [
        {
          "color": "#000",
        },
        {
          "color": "grey",
          "fontSize": 14,
        },
      ]
    }
  >
    john.doe@example.com
  </Text>
  <Text
    style={
      [
        {
          "color": "#000",
        },
        {
          "color": "grey",
          "fontSize": 14,
        },
      ]
    }
  >
    mock
     (
    mock
    ) | 3
  </Text>
  <Text
    style={
      [
        {
          "color": "#000",
        },
        {
          "color": "grey",
          "fontSize": 14,
        },
      ]
    }
  >
    User Context - 
    1234567890
  </Text>
  <Text
    style={
      [
        {
          "color": "#000",
        },
        {
          "color": "grey",
          "fontSize": 14,
        },
      ]
    }
  >
    CTX - 
  </Text>
  <Text
    style={
      [
        {
          "color": "#000",
        },
        {
          "color": "grey",
          "fontSize": 14,
        },
      ]
    }
  >
    Auth - 
  </Text>
  <Text
    style={
      [
        {
          "color": "#000",
        },
        {
          "color": "grey",
          "fontSize": 14,
        },
      ]
    }
  >
    Language - 
    en
  </Text>
  <Text
    style={
      [
        {
          "color": "#000",
        },
        {
          "color": "grey",
          "fontSize": 14,
        },
      ]
    }
  >
    Project - 
    73JwAXeOEhLXUggpVKK9
  </Text>
</View>
`;

exports[`about test 2`] = `<View />`;
